tests/utils/log_reader.py

```diff
@@ -130,56 +157,67 @@ def _parse_line(self, line):
 
         return metric_name, metric_value
 
-    def get_metrics(self, names, timeout=10):
-        def get():
-            _metrics = dict()
-            for name in names:
-                metric_value = self._current_metrics.get(name)
-                if metric_value is None:
-                    return None
-                _metrics[name] = metric_value
-            return _metrics
-
-        start_time = time.time()
-        timeout_time = start_time + timeout
+    def _handle_new_line(self, line):
+        super(LogMetricReader, self)._handle_new_line(line)
+        name, value = self._parse_line(line)
+        self.current_metrics[name] = value
 
-        while not self._stop_event.is_set():
-            with self._lines_lock:
-                metrics = get()
-            if metrics is not None:
-                return metrics
+    def wait_for_metrics_exist(self, names, timeout=10):
+        timeout_time = time.time() + timeout
+        while True:
+            needed_metrics = dict()
+            for name in names:
+                value = self.current_metrics.get(name)
+                if value is None:
+                    break
+                needed_metrics[name] = value
+            else:
+                return needed_metrics
 
-            if time.time() >= timeout_time:
-                raise ValueError(
-                    "timeout of %s seconds reached while waiting for metrics"
-                    % (timeout)
-                )
+            try:
+                self.wait_for_next_line()
+            except LogReaderTimeoutError():
+                if time.time() >= timeout_time:
+                    raise LogReaderTimeoutError(
+                        "Timeout of %s seconds reached while waiting for metrics"
+                        % (timeout)
+                    )
 
-            time.sleep(0.1)
+            time.sleep(0.01)
 
-    def get_metric(self, name):
-        metrics = self.get_metrics([name])
-        return metrics[name]
+    def wait_for_metrics_change(
+        self, metric_names, previous_values, predicate, timeout=10
+    ):
+        timeout_time = time.time() + timeout
+        while True:
+            try:
+                self.wait_for_next_line()
+            except LogReaderTimeoutError:
+                if time.time() >= timeout_time:
```

Comment:
Perhaps we should throw a different exception here (same in other places where we use the same pattern) so we can differentiate between timeout being hit on wait_for_next_line and global timeout for that method being reached.